runtime: move actions back to near_primitives

nagisa opened the pull request:
Rather than keeping these types inside near-vm-runner where they don't belong conceptually and interfere with limited replayability due to being schemas for serialization on-chain, we move them to near_primitives and expose the necessary interfaces to the `near-vm-runner` via `Externals`.

With this none of the near-vm-runner code is on-chain schema sensitive, which is one less thing to worry about.

“But wait,” you ask “shouldn’t limited-replayability help with not needing to think about what changes are made in the first place?” Yes, indeed it should be that way, but it isn’t entirely clear to me if we don’t or won’t have intentional or accidental functionality where the data is not serialized, but rather deserialized, even without a need to invoke the runtime to execute the code. This might be as simple a tool as something that shows the stored blockchain structure (e.g. near explorer.)

(But more seriously, I have petitioned that it would make sense to have all schemas part of the limited replayability story as well, but that didn't really gain as much traction as I had hoped, so I’m just punting this problem upwards and away from the runtime crates.)

cc  #8197

nikurt commented:
I'm lacking domain knowledge.
@Ekleog please take a look.

nagisa commented:
Oh dear, this was meant to be a draft for the time being.

nagisa commented on runtime/runtime/src/ext.rs:
```diff
@@ -195,4 +217,134 @@ impl<'a> External for RuntimeExt<'a> {
+        })
+    }
+
+    fn distribute_unused_gas(&mut self, unused_gas: Gas) -> GasDistribution {
```
I find this really hard to do, actually! There are a couple of competing requirements:

1. `Externals` needs to be object safe => no generics for things like callback closures;
2. Implementation of `ReceiptManager` does not allow for handing out mutable references to `&Gas` through something like an iterator.
    * `fn Externals::unused_gas_recipients(&'a mut self) -> Box<dyn 'a + Iterator<Item = (&'a mut Gas, &'a mut GasWeight)>>` will not work, cuz the `fn Iterator::next(&'1 mut self)` forces `Gas` to need to live for at most `&'1` too (the lifetime shortening rule). Even though that’s not a problem for `slice::IterMut`? I couldn’t figure it out roflmao.
3. I guess I could just create an intermediate allocation, but it seems like people tried really hard to do something more optimal for this code, even though it does not seem hot.

nagisa replied:
Fresh morning head remembered that a `&mut dyn FnMut(...)` is object safe and is actually pretty convenient API wise.

jakmeier commented on runtime/near-vm-runner/src/logic/logic.rs:
```diff
@@ -2794,15 +2768,27 @@ impl<'a> VMLogic<'a> {
+                break 'distribute_gas;
+            }
+            let mut gas_weight_sum: u128 = 0;
+            self.ext
+                .unused_gas_recipients(&mut |_, weight, _| gas_weight_sum += u128::from(weight.0));
             let unused_gas = self.gas_counter.unused_gas();
-
-            // Spend all remaining gas by distributing it among function calls that specify
-            // a gas weight
-            if let GasDistribution::All = self.receipt_manager.distribute_unused_gas(unused_gas) {
-                self.gas_counter.prepay_gas(unused_gas).unwrap();
+            if gas_weight_sum == 0 || unused_gas == 0 {
+                break 'distribute_gas;
             }
+            let mut distributed = 0;
+            self.ext.unused_gas_recipients(&mut |gas, weight, is_last| {
+                let assign = (unused_gas as u128 * weight.0 as u128 / gas_weight_sum) as u64;
+                *gas += assign;
+                distributed += assign;
+                if is_last {
+                    *gas += unused_gas - distributed;
+                }
+            });
+            self.gas_counter.prepay_gas(unused_gas).unwrap();
```
I wonder, isn't this a bit too fancy for the job? Shouldn't we just move all the logic out of here?

I saw your comment that you want to make gas distribution logic something done by the VM runtime, while handling the outgoing receipts is more of a transaction runtime thing. But seeing the tricks necessary to make this cut, I find it hart to believe it is a clean cut.

Another way of looking at it would be to say the `VMOutcome` should just contain the gas weight and leave it up to the transaction runtime to apply them. This isn't burning any gas, hence it doesn't affect gas profiles. And it certainly doesn't affect the main outcome of the function call, only the newly spawned receipts. With that view, I think gas distribution can be seen as a transaction runtime thing, not a VM thing.

WDYT, would that makes sense? I'm not convinced in either direction yet (distributing in VM or in tx runtime) but in my mind, moving all to the transaction runtime would remove a chunk of complicated code.

jakmeier replied:
Sorry for the delay!

So I tried to look at those tests, I only found `function_call_weight_basic_cases_test` and `function_call_no_weight_refund`. The former calls `function_call_weight_check` 10 times, so there are 10 test cases, but the code only needs to be changes in ``function_call_weight_check`.

I believe all other calls to `compute_outcome_and_distribute_gas` don't test anything related to gas weights. They just want to produce a `VMOutcome` but they don't test anything specific to gas weights.

So have to admit I struggle to understand the WWII scenario you faced.^^ Can you maybe elaborate a bit more?


> > VMOutcome should just contain the gas weight
> 
> In a design where the transaction runtime is responsible for dealing with gas weights, `VMOutcome` or really the contract runtime as a whole, needn’t to think about weights at all, other than passing through information about the queued function call actions (which it already does.)

Ahh, yes you are right. Make me think even more that we should move this logic out of the VM.

nagisa replied:
> Make me think even more that we should move this logic out of the VM.

I guess it doesn’t hurt to try again :shrug: 

> So have to admit I struggle to understand the WWII scenario you faced.^^ Can you maybe elaborate a bit more?

I’m quite sure there were more tests than these whose output was dependent on the gas weights, but I can revisit those :shrug:

nagisa replied:
So, there are two tests that are failing after the most recent changes:

* `attach_unspent_gas_but_burn_all_gas`; and
* `function_call_weight_basic_cases_test`.

I only looked at moving the former to `node-runtime` today. The unfortunate thing about the test being there is that none of the test frameworks (e.g. `RuntimeGroup`) work at a low enough level to enable peeking into the promises/actions generated by the contract. The only two viable entry-points for this test from the transaction runtime are `actions::action_function_call` or `actions::execute_function_call`. These are pretty painful to set up an invocation environment for T_T

Looking at `function_call_weight_basic_cases_test` it is using similarly "low-level" APIs to check for what it is checking. To do the same with `actions::execute_function_call` would require some serious rethinking of how these tests work. But I think there's a more straightforward solution – I could move the gas distribution logic entirely into `ReceiptManager` and have this test as part of its unit tests.

jakmeier replied:
Thanks for elaborating!

> I could move the gas distribution logic entirely into ReceiptManager and have this test as part of its unit tests.

That sounds quite reasonable to me.

But overall, since you have already spent quite some time, let me ask this: Do you think it's worth going that route? Or would you rather merge the original proposal, where gas distribution is split between VM and transaction runtime?
Obviously, since you are much deeper into the code right now, I would trust your judgement much more than my own.

nagisa replied:
My intuition is that it is correct to have this code as part of the transaction runtime, and I suspect that we’re better off removing the test(s) that cannot be expressed rather than keeping the logic itself in a less than ideal state and keep the tests.

jakmeier replied:
ok, sounds good to me

jakmeier reviewed (commented):
I haven't gone through the code  100%, yet. But I'm done for today and the comments I left can probably be discussed independent of the changes I didn't look at yet (I think the remainder is mostly test code)

nagisa commented on runtime/runtime/src/ext.rs:
```diff
@@ -96,6 +103,21 @@ impl<'a> RuntimeExt<'a> {
+        let result = result.unwrap_or_else(|payload| std::panic::resume_unwind(payload));
+        assert!(matches!(temp_receipt_manager, None));
+        result
+    }
```
refactored, added documentation.

nagisa commented on runtime/near-vm-runner/src/tests/rs_contract.rs:
```diff
@@ -260,49 +259,6 @@ fn function_call_weight_contract() -> ContractCode {
 }
 
-#[test]
-fn attach_unspent_gas_but_burn_all_gas() {
```
This test is no longer present.

nagisa commented:
@jakmeier I guess, or a rubber-stamp since this already went through a couple iterations of reviews. I haven’t made any significant changes since the last iterations, only fixing up the tests and such so that now it is ready to land.

jakmeier commented:
Ahh, sorry for the delay in that case! @nagisa 

I hadn't looked through the latest changes in the test code, yet, but yeah otherwise nothing new. Somehow I thought there were more open changes.
